File: ts/packages/core/src/domain/erasure-coding.ts
/**
 * K-of-N erasure coding over GF(2^8) for wire-mesh#36's mailbox-opacity case (the shard half of #34's encrypt-then-shard design): a systematic Reed-Solomon construction where the first `dataShards` shards ARE the padded input split into equal runs, and the remaining shards are parity rows of a Vandermonde-style matrix. Any `dataShards` of `totalShards` reconstruct the input -- the property a single mailbox holder must NOT have (it holds one shard and never the epoch key either).
 *
 * Deliberately hand-written rather than a dependency, the same choice the ECIES construction made: the encode format must be byte-identical across this and the Rust implementation, so both pin the SAME vectors against the documented construction rather than trusting two third-party ports to agree. The math is linear algebra over GF(2^8) with the standard 0x11d polynomial (AES's field), not cryptography.
 *
 * The construction, pinned: pad the input with zero bytes to dataShards * ceil(len / dataShards); any shard index below dataShards is that shard's own equal run; parity shard j (0-based within the parity block) has, at each byte position, the GF dot product of the data runs at that position with the coefficient row (alpha^j, alpha^(j+1), ..., alpha^(j+dataShards-1)) where alpha = 0x02 (the field's generator). Decoding: build the dataShards x dataShards submatrix of the full encode matrix selected by the surviving shards' row indices, invert it over the field, and apply it to the surviving shards' bytes. The encode matrix's first dataShards rows are the identity (systematic), so the all-data-shards case needs no inversion at all.
 */

/** The AES field's irreducible polynomial x^8+x^4+x^3+x+1, the conventional GF(2^8) choice. */
const GF_POLY = 0x11d;
/** The bit that appears when a GF(2^8) value overflows a byte and needs the polynomial reduction. */
const GF_OVERFLOW_BIT = 0x100;
/** GF(2^8) caps total shards at 255 (a 256th row would repeat, breaking invertibility). */
const GF_MAX_SHARDS = 255;

export interface ShardConfig {
  dataShards: number;
  totalShards: number;
}

// Plain arrays, not Uint8Array: every element is written before any read (the init loop below), and plain number[] keeps indexed access total under noUncheckedIndexedAccess without a scatter of non-null assertions.
const gfExp: number[] = Array.from({ length: 2 * GF_MAX_SHARDS }, () => 0);
const gfLog: number[] = Array.from({ length: GF_MAX_SHARDS + 1 }, () => 0);
{
  let x = 1;
  for (let power = 0; power < GF_MAX_SHARDS; power += 1) {
    gfExp[power] = x;
    gfLog[x] = power;
    // Multiply by the generator in this field: shift, then reduce by the polynomial when the high bit appears.
    x <<= 1;
    if (x & GF_OVERFLOW_BIT) {
      x ^= GF_POLY;
    }
  }
  // The second half lets gfExp[a+b] index past 255 without a modulo.
  for (let power = GF_MAX_SHARDS; power < 2 * GF_MAX_SHARDS; power += 1) {
    gfExp[power] = gfExp[power - GF_MAX_SHARDS] ?? 0;
  }
}

function gfMul(a: number, b: number): number {
  if (a === 0 || b === 0) return 0;
  return gfExp[(gfLog[a] ?? 0) + (gfLog[b] ?? 0)] ?? 0;
}

function gfInv(a: number): number {
  return gfExp[GF_MAX_SHARDS - (gfLog[a] ?? 0)] ?? 1;
}

function validateConfig(config: Readonly<ShardConfig>): void {
  const { dataShards, totalShards } = config;
  if (!Number.isInteger(dataShards) || dataShards < 1) {
    throw new Error(
      `dataShards must be a positive integer, got ${String(dataShards)}`,
    );
  }
  if (!Number.isInteger(totalShards) || totalShards < dataShards) {
    throw new Error(
      `totalShards must be an integer at least dataShards (${String(dataShards)}), got ${String(totalShards)}`,
    );
  }
  if (totalShards > GF_MAX_SHARDS) {
    throw new Error(
      `the GF(2^8) field supports at most ${String(GF_MAX_SHARDS)} shards, got ${String(totalShards)}`,
    );
  }
}

/** Row i of the encode matrix: identity rows for the data shards, alpha-power rows for parity. */
/**
 * Row i of the encode matrix: identity rows for the data shards, Cauchy rows for parity. Cauchy, not Vandermonde, because only Cauchy guarantees EVERY square submatrix is invertible -- identity-plus-Vandermonde has singular K-subsets (found the hard way: parity-based reconstruction produced wrong bytes on some subsets). With parity row j carrying c_i = 1/(x_j XOR y_i) where x_j = row and y_i = column (all row/column values distinct across the whole matrix, no parity row index equals any data column index, so no denominator is zero), any K surviving rows reduce -- expanding along the identity rows -- to a square Cauchy submatrix, always invertible.
 */
function encodeMatrixRow(row: number, dataShards: number): number[] {
  const coefficients: number[] = Array.from({ length: dataShards }, () => 0);
  if (row < dataShards) {
    coefficients[row] = 1;
    return coefficients;
  }
  for (let column = 0; column < dataShards; column += 1) {
    coefficients[column] = gfInv(row ^ column);
  }
  return coefficients;
}

/** Inverts a square matrix over GF(2^8) by Gauss-Jordan with an identity augmentation. */
function invertMatrix(matrix: readonly number[][]): number[][] {
  const size = matrix.length;
  const work = matrix.map((row, rowIndex) => {
    const augmented: number[] = [
      ...row,
      ...Array.from({ length: size }, () => 0),
    ];
    augmented[size + rowIndex] = 1;
    return augmented;
  });
  for (let column = 0; column < size; column += 1) {
    let pivotRow = -1;
    for (let candidate = column; candidate < size; candidate += 1) {
      if ((work[candidate]?.[column] ?? 0) !== 0) {
        pivotRow = candidate;
        break;
      }
    }
    if (pivotRow === -1) {
      throw new Error("matrix is singular over GF(2^8)");
    }
    if (pivotRow !== column) {
      const swap = work[column] ?? [];
      work[column] = work[pivotRow] ?? [];
      work[pivotRow] = swap;
    }
    const pivotRowValues = work[column];
    if (pivotRowValues === undefined) {
      throw new Error("internal: pivot row vanished");
    }
    const pivotInverse = gfInv(pivotRowValues[column] ?? 1);
    for (let c = 0; c < 2 * size; c += 1) {
      pivotRowValues[c] = gfMul(pivotRowValues[c] ?? 0, pivotInverse);
    }
    for (let other = 0; other < size; other += 1) {
      if (other === column) continue;
      const factor = work[other]?.[column] ?? 0;
      if (factor === 0) continue;
      const target = work[other];
      const source = work[column];
      if (target === undefined || source === undefined) {
        throw new Error("internal: row vanished mid-inversion");
      }
      for (let c = 0; c < 2 * size; c += 1) {
        target[c] = (target[c] ?? 0) ^ gfMul(factor, source[c] ?? 0);
      }
    }
  }
  return work.map((row) => row.slice(size));
}

export async function encodeShards(
  data: Uint8Array,
  config: Readonly<ShardConfig>,
): Promise<Uint8Array<ArrayBuffer>[]> {
  validateConfig(config);
  const { dataShards, totalShards } = config;
  const shardLength = Math.ceil(data.length / dataShards);
  const shards: Uint8Array<ArrayBuffer>[] = [];
  for (let i = 0; i < totalShards; i += 1) {
    shards.push(new Uint8Array(new ArrayBuffer(shardLength)));
  }
  for (let i = 0; i < dataShards; i += 1) {
    shards[i]?.set(data.subarray(i * shardLength, (i + 1) * shardLength));
    // The zero-fill of a fresh Uint8Array IS the padding: the last run simply keeps its trailing zeros per the pinned construction.
  }
  for (
    let parityIndex = dataShards;
    parityIndex < totalShards;
    parityIndex += 1
  ) {
    const coefficients = encodeMatrixRow(parityIndex, dataShards);
    for (let position = 0; position < shardLength; position += 1) {
      let accumulated = 0;
      for (let column = 0; column < dataShards; column += 1) {
        accumulated ^= gfMul(
          coefficients[column] ?? 0,
          shards[column]?.[position] ?? 0,
        );
      }
      const parityTarget = shards[parityIndex];
      if (parityTarget !== undefined) {
        parityTarget[position] = accumulated;
      }
    }
  }
  return Promise.resolve(shards);
}

/**
 * Reconstructs the original bytes from any dataShards of the totalShards
 * produced by encodeShards. originalLength is the manifest's own fact, not
 * derivable from the shards (zero-padding is indistinguishable from a
 * legitimate trailing zero byte), so the caller passes it -- #34's design
 * puts the length in the shard manifest alongside the device/transfer-id list.
 */
/** One surviving shard as presented for reconstruction: its ORIGINAL position among the encode output (the parity equation's row index), never its position in whichever array the caller collected survivors into -- confusing the two silently reconstructs garbage, which is exactly why the index is explicit rather than implied by array order. */
export interface PresentedShard {
  index: number;
  data: Uint8Array;
}

export async function decodeShards(
  presented: readonly PresentedShard[],
  config: Readonly<ShardConfig>,
  originalLength: number,
): Promise<Uint8Array<ArrayBuffer>> {
  validateConfig(config);
  const { dataShards, totalShards } = config;
  for (const entry of presented) {
    if (entry.index < 0 || entry.index >= totalShards) {
      throw new Error(
        `shard index ${String(entry.index)} is outside 0..${String(totalShards - 1)}`,
      );
    }
  }
  if (presented.length < dataShards) {
    throw new Error(
      `need at least ${String(dataShards)} shards to reconstruct, got ${String(presented.length)}`,
    );
  }
  const chosen = presented.slice(0, dataShards);
  const firstChosen = chosen[0];
  if (firstChosen === undefined) {
    throw new Error("internal: chosen shard set is empty");
  }
  const shardLength = firstChosen.data.length;

  // Recover the data shards' full byte content (not just the original prefix): the systematic rows let the all-data case pass through, any other mix goes through the inverted submatrix.
  const recovered: Uint8Array[] = [];
  if (chosen.every((entry) => entry.index < dataShards)) {
    // All data shards survived: identity passthrough, in DATA-shard order
    // (the caller's array order is arbitrary).
    for (let dataIndex = 0; dataIndex < dataShards; dataIndex += 1) {
      const entry = chosen.find((candidate) => candidate.index === dataIndex);
      if (entry === undefined) {
        throw new Error(
          "internal: data shard missing after the all-present check",
        );
      }
      recovered.push(entry.data);
    }
  } else {
    const submatrix: number[][] = chosen.map((entry) =>
      encodeMatrixRow(entry.index, dataShards),
    );
    const inverse = invertMatrix(submatrix);
    for (let dataIndex = 0; dataIndex < dataShards; dataIndex += 1) {
      const row = new Uint8Array(shardLength);
      for (let chosenIndex = 0; chosenIndex < dataShards; chosenIndex += 1) {
        const coefficient = inverse[dataIndex]?.[chosenIndex] ?? 0;
        if (coefficient === 0) continue;
        for (let position = 0; position < shardLength; position += 1) {
          row[position] =
            (row[position] ?? 0) ^
            gfMul(coefficient, chosen[chosenIndex]?.data[position] ?? 0);
        }
      }
      recovered.push(row);
    }
  }

  const totalLength = dataShards * shardLength;
  if (originalLength > totalLength) {
    throw new Error(
      `originalLength ${String(originalLength)} exceeds the shards' capacity ${String(totalLength)} -- wrong manifest for these shards`,
    );
  }
  const stitched = new Uint8Array(new ArrayBuffer(originalLength));
  for (let position = 0; position < originalLength; position += 1) {
    const shardIndex = Math.floor(position / shardLength);
    const within = position % shardLength;
    const byte = recovered[shardIndex]?.[within];
    if (byte === undefined) {
      throw new Error("shard position out of range while stitching");
    }
    stitched[position] = byte;
  }
  return Promise.resolve(stitched);
}

File: rust/crates/wire-mesh-core/src/domain/erasure_coding.rs
//! K-of-N erasure coding over GF(2^8) for wire-mesh#36's mailbox-opacity
//! case (the shard half of #34's encrypt-then-shard design) -- the Rust
//! half of the byte-identical pair with
//! `ts/packages/core/src/domain/erasure-coding.ts`. A systematic
//! construction: the first `data_shards` shards ARE the zero-padded input
//! split into equal runs, and the remaining shards are Cauchy parity rows
//! `c_i = 1/(x_j XOR y_i)` with `x_j = row`, `y_i = column`. Cauchy, not
//! Vandermonde, because only Cauchy guarantees EVERY square submatrix is
//! invertible -- identity-plus-Vandermonde has singular K-subsets. Any
//! `data_shards` of `total_shards` reconstruct the input; a single mailbox
//! holder has one shard and never the epoch key either.
//!
//! The pinned cross-language vector (asserted identically in both suites):
//! K=2 N=3 over "wire-mesh" produces shards "wire-", "mesh\0", and the
//! parity bytes `[101, 153, 227, 111, 152]`.

/// The AES field's irreducible polynomial x^8+x^4+x^3+x+1, the conventional GF(2^8) choice.
const GF_POLY: u16 = 0x11d;
/// The bit that appears when a GF(2^8) value overflows a byte and needs the polynomial reduction.
const GF_OVERFLOW_BIT: u16 = 0x100;
/// GF(2^8) caps total shards at 255 (a 256th row would repeat, breaking invertibility).
const GF_MAX_SHARDS: usize = 255;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct ShardConfig {
    pub data_shards: usize,
    pub total_shards: usize,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ErasureError {
    /// data_shards < 1, total_shards < data_shards, or shards beyond the field.
    InvalidConfig(String),
    /// Fewer than data_shards surviving shards were presented.
    TooFewShards { have: usize, need: usize },
    /// A presented shard index is outside 0..total_shards.
    IndexOutOfRange { index: usize },
    /// The manifest's original length exceeds the shards' padded capacity.
    LengthExceedsCapacity { length: usize, capacity: usize },
}

impl core::fmt::Display for ErasureError {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        match self {
            ErasureError::InvalidConfig(what) => write!(f, "invalid shard config: {what}"),
            ErasureError::TooFewShards { have, need } => {
                write!(f, "need at least {need} shards to reconstruct, got {have}")
            }
            ErasureError::IndexOutOfRange { index } => {
                write!(f, "shard index {index} is outside the configured range")
            }
            ErasureError::LengthExceedsCapacity { length, capacity } => write!(
                f,
                "original length {length} exceeds the shards' capacity {capacity} -- wrong manifest for these shards"
            ),
        }
    }
}

impl std::error::Error for ErasureError {}

/// Exponent and log tables over the field, built once at first use.
struct Tables {
    exp: Vec<u8>,
    log: Vec<u8>,
}

fn tables() -> &'static Tables {
    use std::sync::OnceLock;
    static TABLES: OnceLock<Tables> = OnceLock::new();
    TABLES.get_or_init(|| {
        let mut exp = vec![0u8; 2 * GF_MAX_SHARDS];
        let mut log = vec![0u8; GF_MAX_SHARDS + 1];
        let mut x: u16 = 1;
        for (power, slot) in exp.iter_mut().enumerate().take(GF_MAX_SHARDS) {
            *slot = x as u8;
            log[x as usize] = power as u8;
            x <<= 1;
            if x & GF_OVERFLOW_BIT != 0 {
                x ^= GF_POLY;
            }
        }
        // The second half lets exp[a+b] index past 255 without a modulo.
        for power in GF_MAX_SHARDS..2 * GF_MAX_SHARDS {
            exp[power] = exp[power - GF_MAX_SHARDS];
        }
        Tables { exp, log }
    })
}

fn gf_mul(a: u8, b: u8) -> u8 {
    if a == 0 || b == 0 {
        return 0;
    }
    let t = tables();
    t.exp[(t.log[a as usize] as usize) + (t.log[b as usize] as usize)]
}

fn gf_inv(a: u8) -> u8 {
    let t = tables();
    t.exp[GF_MAX_SHARDS - t.log[a as usize] as usize]
}

fn validate_config(config: ShardConfig) -> Result<(), ErasureError> {
    let ShardConfig {
        data_shards,
        total_shards,
    } = config;
    if data_shards < 1 {
        return Err(ErasureError::InvalidConfig(format!(
            "data_shards must be at least 1, got {data_shards}"
        )));
    }
    if total_shards < data_shards {
        return Err(ErasureError::InvalidConfig(format!(
            "total_shards must be at least data_shards ({data_shards}), got {total_shards}"
        )));
    }
    if total_shards > GF_MAX_SHARDS {
        return Err(ErasureError::InvalidConfig(format!(
            "the GF(2^8) field supports at most {GF_MAX_SHARDS} shards, got {total_shards}"
        )));
    }
    Ok(())
}

/// Row `row` of the encode matrix: identity rows for the data shards,
/// Cauchy rows for parity (see the module doc for the guarantee).
fn encode_matrix_row(row: usize, data_shards: usize) -> Vec<u8> {
    let mut coefficients = vec![0u8; data_shards];
    if row < data_shards {
        coefficients[row] = 1;
        return coefficients;
    }
    for (column, slot) in coefficients.iter_mut().enumerate() {
        *slot = gf_inv((row ^ column) as u8);
    }
    coefficients
}

/// Inverts a square matrix over GF(2^8) by Gauss-Jordan with an identity
/// augmentation. Errors (rather than panics) on a singular input -- for
/// the Cauchy construction that is unreachable, but this is generic math.
fn invert_matrix(matrix: &[Vec<u8>]) -> Result<Vec<Vec<u8>>, ErasureError> {
    let size = matrix.len();
    let mut work: Vec<Vec<u8>> = matrix
        .iter()
        .enumerate()
        .map(|(row_index, row)| {
            let mut augmented = row.clone();
            augmented.resize(2 * size, 0);
            augmented[size + row_index] = 1;
            augmented
        })
        .collect();
    for column in 0..size {
        let pivot_row = (column..size).find(|candidate| work[*candidate][column] != 0);
        let pivot_row = match pivot_row {
            Some(found) => found,
            None => {
                return Err(ErasureError::InvalidConfig(
                    "matrix is singular over GF(2^8)".to_owned(),
                ))
            }
        };
        if pivot_row != column {
            work.swap(column, pivot_row);
        }
        let pivot_inverse = gf_inv(work[column][column]);
        for value in work[column].iter_mut() {
            *value = gf_mul(*value, pivot_inverse);
        }
        for other in 0..size {
            if other == column {
                continue;
            }
            let factor = work[other][column];
            if factor == 0 {
                continue;
            }
            let source = work[column].clone();
            let target = &mut work[other];
            for (target_slot, source_slot) in target.iter_mut().zip(source.iter()) {
                *target_slot ^= gf_mul(factor, *source_slot);
            }
        }
    }
    Ok(work.into_iter().map(|row| row[size..].to_vec()).collect())
}

/// Splits `data` into `total_shards` shards, any `data_shards` of which
/// reconstruct it: the first `data_shards` are the zero-padded input in
/// equal runs (systematic), the rest are Cauchy parity.
pub fn encode_shards(data: &[u8], config: ShardConfig) -> Result<Vec<Vec<u8>>, ErasureError> {
    validate_config(config)?;
    let ShardConfig {
        data_shards,
        total_shards,
    } = config;
    let shard_length = data.len().div_ceil(data_shards);
    let mut shards = vec![vec![0u8; shard_length]; total_shards];
    for (i, shard) in shards.iter_mut().enumerate().take(data_shards) {
        // The last run may be shorter than shard_length (the input is not a
        // multiple of K): copy what exists, and the zero-fill of a fresh
        // shard IS the padding for the rest.
        let start = i * shard_length;
        let end = ((i + 1) * shard_length).min(data.len());
        shard[..end - start].copy_from_slice(&data[start..end]);
    }
    for parity_index in data_shards..total_shards {
        let coefficients = encode_matrix_row(parity_index, data_shards);
        let parity_bytes: Vec<u8> = (0..shard_length)
            .map(|position| {
                let mut accumulated: u8 = 0;
                for (column, coefficient) in coefficients.iter().enumerate() {
                    accumulated ^= gf_mul(*coefficient, shards[column][position]);
                }
                accumulated
            })
            .collect();
        shards[parity_index] = parity_bytes;
    }
    Ok(shards)
}

/// One surviving shard as presented for reconstruction: its ORIGINAL
/// position among the encode output (the parity equation's row index),
/// never its position in whichever collection the caller gathered
/// survivors into.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PresentedShard {
    pub index: usize,
    pub data: Vec<u8>,
}

/// Reconstructs the original bytes from any `data_shards` of the
/// `total_shards` produced by [`encode_shards`]. `original_length` is the
/// manifest's own fact -- zero-padding is indistinguishable from a
/// legitimate trailing zero byte, so the length is not derivable from the
/// shards themselves.
pub fn decode_shards(
    presented: &[PresentedShard],
    config: ShardConfig,
    original_length: usize,
) -> Result<Vec<u8>, ErasureError> {
    validate_config(config)?;
    let ShardConfig {
        data_shards,
        total_shards,
    } = config;
    for entry in presented {
        if entry.index >= total_shards {
            return Err(ErasureError::IndexOutOfRange { index: entry.index });
        }
    }
    if presented.len() < data_shards {
        return Err(ErasureError::TooFewShards {
            have: presented.len(),
            need: data_shards,
        });
    }
    let chosen = &presented[..data_shards];
    let shard_length = chosen[0].data.len();

    // Recover every data shard's full padded content. All-data-shards is an
    // identity passthrough (in DATA-shard order, not caller order); any
    // other mix goes through the inverted submatrix.
    let mut recovered: Vec<Vec<u8>> = Vec::with_capacity(data_shards);
    if chosen.iter().all(|entry| entry.index < data_shards) {
        for data_index in 0..data_shards {
            let entry = chosen
                .iter()
                .find(|candidate| candidate.index == data_index)
                .ok_or_else(|| {
                    ErasureError::InvalidConfig(
                        "internal: data shard missing after the all-present check".to_owned(),
                    )
                })?;
            recovered.push(entry.data.clone());
        }
    } else {
        let submatrix: Vec<Vec<u8>> = chosen
            .iter()
            .map(|entry| encode_matrix_row(entry.index, data_shards))
            .collect();
        let inverse = invert_matrix(&submatrix)?;
        for inverse_row in inverse.iter().take(data_shards) {
            let mut row = vec![0u8; shard_length];
            for (chosen_index, coefficient) in inverse_row.iter().enumerate() {
                if *coefficient == 0 {
                    continue;
                }
                let source = &chosen[chosen_index].data;
                for (position, slot) in row.iter_mut().enumerate() {
                    let contribution = gf_mul(*coefficient, source[position]);
                    *slot ^= contribution;
                }
            }
            recovered.push(row);
        }
    }

    let capacity = data_shards * shard_length;
    if original_length > capacity {
        return Err(ErasureError::LengthExceedsCapacity {
            length: original_length,
            capacity,
        });
    }
    let mut stitched = vec![0u8; original_length];
    for (position, out) in stitched.iter_mut().enumerate() {
        let shard_index = position / shard_length;
        let within = position % shard_length;
        *out = recovered[shard_index][within];
    }
    Ok(stitched)
}

#[cfg(test)]
mod tests {
    use super::*;

    const K2_N2: ShardConfig = ShardConfig {
        data_shards: 2,
        total_shards: 2,
    };
    const K2_N3: ShardConfig = ShardConfig {
        data_shards: 2,
        total_shards: 3,
    };
    const K2_N4: ShardConfig = ShardConfig {
        data_shards: 2,
        total_shards: 4,
    };
    const K2_N5: ShardConfig = ShardConfig {
        data_shards: 2,
        total_shards: 5,
    };
    const K3_N4: ShardConfig = ShardConfig {
        data_shards: 3,
        total_shards: 4,
    };
    const K3_N5: ShardConfig = ShardConfig {
        data_shards: 3,
        total_shards: 5,
    };

    fn text(value: &str) -> Vec<u8> {
        value.as_bytes().to_vec()
    }

    fn survivors_of(shards: &[Vec<u8>], dropped: &[usize]) -> Vec<PresentedShard> {
        shards
            .iter()
            .enumerate()
            .filter(|(index, _)| !dropped.contains(index))
            .map(|(index, data)| PresentedShard {
                index,
                data: data.clone(),
            })
            .collect()
    }

    #[test]
    fn reconstructs_from_every_k_subset() {
        let data = text("erasure-coded notice content");
        let shards = encode_shards(&data, K3_N5).expect("encode");
        assert_eq!(shards.len(), K3_N5.total_shards);
        for drop1 in 0..K3_N5.total_shards {
            for drop2 in (drop1 + 1)..K3_N5.total_shards {
                let decoded =
                    decode_shards(&survivors_of(&shards, &[drop1, drop2]), K3_N5, data.len())
                        .expect("decode");
                assert_eq!(decoded, data);
            }
        }
    }

    #[test]
    fn round_trips_k_equals_n_pure_splitting() {
        let data = text("split me");
        let shards = encode_shards(&data, K2_N2).expect("encode");
        assert_eq!(shards[0], text("spli"));
        assert_eq!(shards[1], text("t me"));
        let decoded =
            decode_shards(&survivors_of(&shards, &[]), K2_N2, data.len()).expect("decode");
        assert_eq!(decoded, data);
    }

    #[test]
    fn reconstructs_from_first_k_shards() {
        let data = text("data shards alone");
        let shards = encode_shards(&data, K3_N4).expect("encode");
        let only_parity_dropped = [K3_N4.data_shards];
        let decoded = decode_shards(
            &survivors_of(&shards, &only_parity_dropped),
            K3_N4,
            data.len(),
        )
        .expect("decode");
        assert_eq!(decoded, data);
    }

    #[test]
    fn reconstructs_with_zero_data_shards_parity_only() {
        let data = text("parity only");
        let shards = encode_shards(&data, K2_N5).expect("encode");
        let decoded =
            decode_shards(&survivors_of(&shards, &[0, 1]), K2_N5, data.len()).expect("decode");
        assert_eq!(decoded, data);
    }

    #[test]
    fn pads_the_last_shard_and_trims_to_original_length() {
        let data = text("abcde");
        let shards = encode_shards(&data, K3_N5).expect("encode");
        let decoded =
            decode_shards(&survivors_of(&shards, &[1, 2]), K3_N5, data.len()).expect("decode");
        assert_eq!(decoded, data);
        assert_eq!(decoded.len(), data.len());
    }

    #[test]
    fn rejects_fewer_than_k_shards() {
        let data = text("too few");
        let shards = encode_shards(&data, K3_N5).expect("encode");
        let result = decode_shards(&survivors_of(&shards, &[2, 3, 4]), K3_N5, data.len());
        assert!(matches!(
            result,
            Err(ErasureError::TooFewShards { have: 2, need: 3 })
        ));
    }

    #[test]
    fn rejects_impossible_configs() {
        let data = text("x");
        assert!(matches!(
            encode_shards(
                &data,
                ShardConfig {
                    data_shards: 3,
                    total_shards: 2
                }
            ),
            Err(ErasureError::InvalidConfig(_))
        ));
        assert!(matches!(
            encode_shards(
                &data,
                ShardConfig {
                    data_shards: 256,
                    total_shards: 257
                }
            ),
            Err(ErasureError::InvalidConfig(_))
        ));
    }

    #[test]
    fn rejects_shard_index_out_of_range() {
        let data = text("bounds");
        let shards = encode_shards(&data, K2_N3).expect("encode");
        let presented = vec![
            PresentedShard {
                index: 0,
                data: shards[0].clone(),
            },
            PresentedShard {
                index: 7,
                data: shards[1].clone(),
            },
        ];
        assert!(matches!(
            decode_shards(&presented, K2_N3, data.len()),
            Err(ErasureError::IndexOutOfRange { index: 7 })
        ));
    }

    #[test]
    fn empty_input_round_trips() {
        let shards = encode_shards(&[], K2_N3).expect("encode");
        let decoded = decode_shards(&survivors_of(&shards, &[1]), K2_N3, 0).expect("decode");
        assert!(decoded.is_empty());
    }

    #[test]
    fn handles_binary_input_including_zero_bytes() {
        let data = vec![0u8, 255, 0, 1, 0, 128, 0, 0, 7];
        let shards = encode_shards(&data, K2_N4).expect("encode");
        let drop_sets: [&[usize]; 8] = [&[0], &[1], &[2], &[3], &[0, 1], &[0, 2], &[1, 3], &[2, 3]];
        for dropped in drop_sets {
            let decoded =
                decode_shards(&survivors_of(&shards, dropped), K2_N4, data.len()).expect("decode");
            assert_eq!(decoded, data);
        }
    }

    /// The cross-language pin: byte-identical to the TypeScript suite's own
    /// assertion over the same input and configuration.
    #[test]
    fn produces_the_byte_pinned_cross_language_vector() {
        let data = text("wire-mesh");
        let shards = encode_shards(&data, K2_N3).expect("encode");
        assert_eq!(shards[0], text("wire-"));
        assert_eq!(shards[1], {
            let mut padded = text("mesh");
            padded.push(0);
            padded
        });
        assert_eq!(shards[2], vec![101, 153, 227, 111, 152]);
        let presented = vec![
            PresentedShard {
                index: 0,
                data: shards[0].clone(),
            },
            PresentedShard {
                index: 2,
                data: shards[2].clone(),
            },
        ];
        let round_trip = decode_shards(&presented, K2_N3, data.len()).expect("decode");
        assert_eq!(round_trip, data);
    }
}
